### ISSUE TITLE
Google Search Indexing - Video is not the main content of the page

### ISSUE THREAD

**Issue by DVDGuy99** [Chocobozzz/PeerTube#6210]
### Describe the current behavior

Google Search Console gives the error "Video is not the main content of the page" when indexing videos on our PeerTube site.

This is one of the video pages that Google says the video is not the main content:

[https://trailers.ddigest.com/w/sbKiNjKTCs8EkNpKb45ku9](https://trailers.ddigest.com/w/f7wYPrAQjheqnn2RaZbkPU)

(actually, I think it will say that for all the video pages - diving deeper into the video page indexing data, it says « Video is supplementary content on the page »)

Possibly related, but viewing a screenshot of the page generated within Google Search Console shows it displaying a "HLS.js does not seem to be supported" error where the video should be. Doing an exact term search for this on the videos section of Google shows quite a few PeerTube-hosted videos that have this as the crawled text description for the video.

### Steps to reproduce

1. Log in to Google Search Console account for the instance
2. Under "Indexing" go to "Video Pages"
3. A list of pages with the "Video is not the main content of the page" error should be listed here

### Describe the expected behavior

As these pages are the main video playback pages, the video should of course be the main content of the page and Google should index these as such. Pages with videos that are indexed as the main content will show the video carousel with the video thumbnail as opposed to just a text link.

### Additional information

* PeerTube instance:
  * URL: https://trailers.ddigest.com/
  * Version: 6.0.3
  * NodeJS version: 
  * Ffmpeg version:

* Browser name, version and platforms on which you could reproduce the bug:
* Link to browser console log if relevant:
* Link to server log if relevant (`journalctl` or `/var/www/peertube/storage/logs/`):

**Comment by Chocobozzz** [comment]
Google bot seems to fail to load the HLS player (which is a non-sense). Trying to fallback to raw HTML element using https://github.com/Chocobozzz/PeerTube/commit/c4a062109d562cbe505c17044dd0b569a92ea121

Hope it will fix the issue (have to wait deploy on peertube2.cpy.re and re-schedule a google bot indexation)

**Comment by Chocobozzz** [comment]
Seems like to fix the issue :+1:

**Comment by aflamrip** [comment]
Has the problem been solved or is the same problem still present?
![Indexing-pages-with-videos-URL-inspection](https://github.com/Chocobozzz/PeerTube/assets/112701602/8f55f6e1-8d14-497a-931f-7e0c73579938)
![1g](https://github.com/Chocobozzz/PeerTube/assets/112701602/c82ea405-9fc4-43bb-adb3-f4a1c85c9b5c)

**Comment by Chocobozzz** [comment]
> Has the problem been solved or is the same problem still present?

Should be fixed in next peertube release (6.1.0)

**Comment by aflamrip** [comment]
I think I found a solution but I don't know if it is right or wrong
On this path
peertube-latest\client\dist\standalone\videos\
There is a file
embed.html
This part is modified
<meta name="robots" content="noindex">
to
<meta name="robots" content="index">

But I don't know if this method will solve the problem
Video is not the main content of the page

Video placement

Video is supplementary content on the page

Whether the page is a playback page for a single video (Video is main content on the page), or hosts additional meaningful content or videos (Video is supplementary content on the page).

**Comment by DVDGuy99** [comment]
This issue seems to be still present in 6.1.0.

I've tried enabling/disabling web video, HLS with P2P support, and it doesn't seem to matter too much, as it still gives the "video is not the main content of the page" error:

<img width="715" alt="google_search_console_not_main_content" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/55aec8da-3851-4904-9f40-59597aff1719">

Below is the JavaScript console error messages as shown in Google Search Console for a sample page (https://trailers.ddigest.com/w/1ZcXuBacku4tZeY7KPHwPF), including it in case it helps:

<img width="374" alt="google_search_console_errors" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/b47711b5-f396-432d-bf7c-80089d83954f">

**Comment by DVDGuy99** [comment]
Here's the video page indexing report for another video with web video enabled:

<img width="671" alt="google_search_console_not_main_content2" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/327c83e2-53df-409b-960a-1ee3c06ec1c4">

**Comment by Chocobozzz** [comment]
It's a nonsense, sometimes Google considers the video is not the main content on the page, and a few days later it correctly indexes the video. I'll look into it again, but if anyone here has a any clue, here don't hesitate to share it

**Comment by DVDGuy99** [comment]
This thread might shed some light, and I think there's a really stupid fix for all of this involving adding the word "video" to the URL:

https://support.google.com/webmasters/thread/247936417/how-to-fix-video-is-not-the-main-content-of-the-page?hl=en

There are only 4 videos on my site that have been indexed and the URL that is indexed is like this:

https://trailers.ddigest.com/videos/watch/218beda6-427d-4ba5-83ad-d815cd13fbc6

Whereas all the ones not indexed is like this:

https://trailers.ddigest.com/w/jtdUAPbo65bNgz4Momxmm4

I wonder if a separate Google sitemap can be created that uses the "video/watch" URL structure as opposed to the "w/" one. For now, Google doesn't seem to care if the first one redirects to the second one.

**Comment by DVDGuy99** [comment]
I've set up a cron job to create a version of the sitemap to be a workaround for this issue. The script basically replaces "https://trailers.ddigest.com/w/" with "https://trailers.ddigest.com/videos/watch/" in the sitemap, and then replaced the submitted sitemap in Google Search Console with this newly edited sitemap. This seems to work and videos are now being indexed, even though it shouldn't (as I'm submitting pages with redirects):

<img width="710" alt="Screenshot 2024-06-01 141625" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/95524670-8c3a-4c85-ad43-ae98b18ad1e0">

**Comment by Chocobozzz** [comment]
@DVDGuy99 Coming to the news: do google index all your videos with the new `/videos/watch` now?

**Comment by DVDGuy99** [comment]
@Chocobozzz Yes, pretty much. It doesn't seem to re-add the videos that have already been indexed, even if they've been submitted via the sitemap. I'll try to force the reindexing (via the request indexing feature in Google Search Consoles) on a couple of older ones to see if they are also added/re-added.

**Comment by DVDGuy99** [comment]
The older videos I've requested reindexing for have also been indexed as videos (moved out of the "Video is not the main content of the page" category), as have all the new videos that are included in my modified sitemap.

**Comment by Chocobozzz** [comment]
Unfortunately it doesn't work on my side, indexing https://framatube.org/videos/watch/gW6BUFLNSDWWZwUzZBXLoN instead of https://framatube.org/w/gW6BUFLNSDWWZwUzZBXLoN is refused by google because of the redirection. I'm surprised it works on your instance :thinking:

**Comment by DVDGuy99** [comment]
It's definitely a weird situation with Google at the moment, and I'm thinking it has to be a bug or something. I have several videos where, as you said, the page won't get indexed because it's a redirect, but the video (with the "/videos/watch/" URL) does get indexed. So it seems that video pages only get indexed if it has "video" in the URL even if it's a redirect. I'm going to submit the "/w/" version of the URL for these pages and see what happens - maybe the page gets indexed but the video indexing is removed due to the "not the main content" error.

**Comment by DVDGuy99** [comment]
So I managed to get both versions of the page indexed by Google. Don't ask me how it works, it's not supposed to, but it does for me.

<img width="890" alt="Screenshot 2024-07-06 160014" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/74bcb1e1-9a05-4843-b361-039ef151284c">
<img width="899" alt="Screenshot 2024-07-06 155950" src="https://github.com/Chocobozzz/PeerTube/assets/159438367/103b57e3-ba79-437e-b2ff-d9b4e3a9d282">

**Comment by kontrollanten** [comment]
According to [Googles docs](https://developers.google.com/search/docs/crawling-indexing/sitemaps/video-sitemaps) it's recommended to have a `video:content_loc` tag in the sitemap, which currently doesn't exist in Peertubes sitemap.

> It's required to provide either a <video:content_loc> or <video:player_loc> tag. We recommend that your provide the <video:content_loc> tag, if possible. This is the most effective way for Google to fetch your video content files. If <video:content_loc> isn't available, provide <video:player_loc> as an alternative.

Another explanation may be those client logs whom seems to come from Googlebot:

```
{
    "tags": [
        "client"
    ],
    "userAgent": "Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.6613.113 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)",
    "meta": "{\"currentTime\":0,\"data\":{\"type\":\"mediaError\",\"details\":\"manifestIncompatibleCodecsError\",\"fatal\":true,\"url\":\"https://cdn.peertube/streaming-playlists-native/hls/4bbb3e31-24fa-4ec4-8daa-ec6f1d54b4ef/bab5b5b4-884e-456e-9a84-fcd2f6a6e623-master.m3u8\",\"error\":{},\"reason\":\"no level with compatible codecs found in manifest\"}}",
    "url": "https://peertube/w/4bbb3e31-24fa-4ec4-8daa-ec6f1d54b4ef",
    "level": "error",
    "message": "Client log: HLS.js error: mediaError - fatal: true - manifestIncompatibleCodecsError",
    "timestamp": "2024-09-11T11:58:06.448Z"
}
```

**Comment by Chocobozzz** [comment]
> According to [Googles docs](https://developers.google.com/search/docs/crawling-indexing/sitemaps/video-sitemaps) it's recommended to have a video:content_loc tag in the sitemap, which currently doesn't exist in Peertubes sitemap.

Why not, but I think most other web video platforms (youtube, vimeo...) don't include this tag but are still indexed :angry: 

> Another explanation may be those client logs whom seems to come from Googlebot:

I think it's an expected behaviour where Googlebot disabled video support in its engine.

**Comment by kontrollanten** [comment]
> Why not, but I think most other web video platforms (youtube, vimeo...) don't include this tag but are still indexed 😠

Sure, but I think Googlebot makes some kind of holistic assessment where other platforms has higher general ranking, loads faster, is easier to crawl, etc. So if we try to perfect on all points, maybe it'll be indexed.

**Comment by Chocobozzz** [comment]
Seems the issue is mostly fixed on our peertube instances. @kontrollanten, @DVDGuy99 can you confirm?

**Comment by kontrollanten** [comment]
We haven't installed PT 7 yet

**Comment by DVDGuy99** [comment]
Most of our new videos are still being put into the "video is not the main content" category (it has been renamed "Video isn't on watch page" by Google). But I do find that after a while (several weeks to more than a month), some of these will be automatically categorised as being video pages. Right now, 60% of our video pages are indexed as videos, the 40% (including almost all the recently uploaded ones) is in the "Video isn't on watch page" category.

I've requested a fix validation by Google to see if this can clear up some of these pages, will post the results here.

**Comment by DVDGuy99** [comment]
The validation eventually failed, as the same error (now called "Video isn't on a watch page") still occurs. However, I do find over time some of these pages will become classified as video pages automatically. Here's Google's document on what is considered a "watch page":

https://developers.google.com/search/docs/appearance/video?sjid=7178086079256790892-NC#watch-page

One of my theories is that it's also a possibility that pages that have low visitor rates and incoming links may find it hard to be classified as a watch page. This is why when pages get more links and views, these then automatically get re-classified as a watch page by Google at a much later date.

**Comment by Chocobozzz** [comment]
> One of my theories is that it's also a possibility that pages that have low visitor rates and incoming links may find it hard to be classified as a watch page. This is why when pages get more links and views, these then automatically get re-classified as a watch page by Google at a much later date.

I agree with this theory. I'm closing this issue but don't hesitate to comment it if you think we can improve things on PeerTube side, or find other interesting info 👍

**Comment by kontrollanten** [comment]
We've upgraded to v7.1.0 and asked google for a reindex. The reindexation failed to verify that it was correct though, it still fails due to "Video is not the main content". It gives no further information.

**Comment by aflamrip** [comment]
The problem has reappeared again.
Video isn't on a watch page.

I think the problem may be caused by the "Other Videos" menu on the watch page.
What I found when the webmasters checked one of the links was that the number of videos detected was:
One video was detected on the page.

However, in the section titled "Why wasn't the video indexed?", this appears:
"The video is not included on the watch page."
"Does it appear on the watch page?"
"No, the video is supplementary content on the page."

The reason:
If the primary purpose of the page is to display a single video to users, it is considered a watch page. However, if the page includes additional useful content or videos, it is not considered a watch page.

Therefore, the problem may be caused by the "Other Videos" menu on the watch page.

Regarding this section:
If the page includes additional useful content or videos, it is not considered a watch page.

The strange thing is that some videos actually get re-categorized over time and appear in video search for no apparent reason.

**Comment by kontrollanten** [comment]
This may be related to https://github.com/videojs/video.js/issues/9171

**Comment by kontrollanten** [comment]
> This may be related to [videojs/video.js#9171](https://github.com/videojs/video.js/issues/9171)

Now released in 8.24.0

**Comment by Chocobozzz** [comment]
> Now released in 8.24.0

Thank you, implemented in https://github.com/Chocobozzz/PeerTube/commit/bedeb18cc3255a827285096a3b449b350448ca68

I also tried to fix this SEO issue in the following commits:
 * https://github.com/Chocobozzz/PeerTube/commit/0fbf9ed910669379298a36b8685ef5e4465bc8e3
 * https://github.com/Chocobozzz/PeerTube/commit/41db7da56424297b83108e76442952a67bf47653

**Comment by kontrollanten** [comment]
Thanks! We deployed the commit to our production environment. As anonymous user the video poster is still lazy loaded though.

**Comment by Chocobozzz** [comment]
I think it's because the video is not auto played because of web browser blocking autoplay? I suggest waiting to see if these changes improves this issue, before eagerly load the poster for everyone even if the video is auto played.
